### PR TITLE
Temporal min points

### PR DESCRIPTION
**Pull request by gennartan**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | (#4364) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Tested using colcon test only|
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

* Add 2 new parmeters `min_consecutive_in` and `min_consecutive_out` that adds a temporal axis to the `min_points` parameter.

## Description of documentation updates required from your changes

* Added 2 new parameters, so need to add that to default configs and documentation page

---

## Future work that may be required in bullet points

* The `min_consecutive_out` parameter does not work for the `PolygonApproach`  type because the collision time is computed with the latest sensor measurements. So whenever a measurement is outside the range, the collision time will be `-1` and we will not be able to keep the robot at a slower speed. So it would be required to remember the previous `collision_time` and take an appropriate decision (keep the worst, keep the best, ... ?) 

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Make sure to pull in main so that CI works for the Rotation Shim fix

File: nav2_collision_monitor/src/collision_monitor_node.cpp [hunk 1]
```diff
@@ -571,7 +573,8 @@ bool CollisionMonitor::processApproach(
 
   // Obtain time before a collision
   const double collision_time = polygon->getCollisionTime(collision_points, velocity);
-  if (collision_time >= 0.0) {
+  polygon->newDecisionPoint(collision_time >= 0.0);
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/collision_monitor_node.cpp:576]: I don't think this is sensible for the approach polygon, no?

File: nav2_collision_monitor/src/polygon.cpp [hunk 2]
```diff
@@ -229,6 +230,28 @@ int Polygon::getPointsInside(const std::vector<Point> & points) const
   return num;
 }
 
+void Polygon::newDecisionPoint(bool is_in)
+{
+  if (is_in && current_consecutive_in_ < min_consecutive_in_) {
+    current_consecutive_in_++;
+    current_consecutive_out_ = 0;
+    if (current_consecutive_in_ >= min_consecutive_in_) {
+      current_decision_ = true;
```
Comment (SteveMacenski) [inline nav2_collision_monitor/src/polygon.cpp:239]: I think the current state needs to be reset on `activate` so that you can activate / deactivate and have it work without stale measurements

File: nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp [hunk 3]
```diff
@@ -145,6 +145,19 @@ class Polygon
    */
   virtual int getPointsInside(const std::vector<Point> & points) const;
 
+  /**
+   * @brief Add a new decision point to count the consecutive number of
+   * time enough point are in the polygon. Updates the current decision value
+   * based on previous measurements.
+   */
+  void newDecisionPoint(bool is_in);
```
Comment (SteveMacenski) [inline nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:153]: On naming: this is a little ambiguous as to the what decision it is. Can you come up with some alternatives?

File: nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp [hunk 4]
```diff
@@ -251,6 +264,10 @@ class Polygon
   ActionType action_type_;
   /// @brief Minimum number of data readings within a zone to trigger the action
   int min_points_;
+  /// @brief Minimum number of consecutive data with minimum min_points to trigger the action
+  int min_consecutive_in_;
```
Comment (SteveMacenski) [inline nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:268]: On naming: this a little ambiguous as a variable / parameter name. Can you come up with some alternatives?

File: nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp [hunk 5]
```diff
@@ -251,6 +264,10 @@ class Polygon
...
+  /// @brief Minimum number of consecutive data without minimum min_points to stop the action
+  int min_consecutive_out_;
```
Comment (SteveMacenski) [inline nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:270]: Thoughts about a duration rather than a number of consecutive cycles? I would think that having the system blocked for > 0.5 seconds or something is more intuitive than 5 cycles, or what have you. I think that would be more intuitive to me, but open to leaving as is if others prefer this (@tonynajjar ?)

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp (hunk 5 above)
Comment (tonynajjar) [reply nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:270]: I agree that specifying a duration would be more intuitive especially because the CM and CD's run on a rate different than the rate of the measurements. For CD it's a configurable rate and for CM it's the rate of the cmd_vel. The "meaning" of `min_consecutive_out_` is dependent on the difference between the CM/CD rate and the measurement rate, if that makes sense.

TLDR: I would vote for duration

Comment (SteveMacenski) [reply nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:270]: @gennartan what do you think?

Comment (gennartan) [reply nav2_collision_monitor/include/nav2_collision_monitor/polygon.hpp:270]: Specifying a duration is a better idea, I will update the PR soon and come back with better names for the variables

**Comment by SteveMacenski** [comment]
@gennartan I see that you pushed some updates, should I take a look and review or just incremental updates? 

CI failed ... oddly, might be worth rebasing your PR, but I retriggered CI to see if it was just a fluke. Note the linting test failures from `ament_cpplint` and `ament_uncrustify`, those are important

**Comment by gennartan** [comment]
Hi, just incremental updates for now, I still need to add some testing to it. I will check why the tests are failling and fix that.

Do you know if there is a method to create tests with precise timing to simulate the clock in ROS ?

**Comment by SteveMacenski** [comment]
Ok!

> Do you know if there is a method to create tests with precise timing to simulate the clock in ROS ?

I think an illustrative example of what you’re looking to do might help me give a suggestion. I’m not sure that should be required here, but maybe we’re thinking about different definitions of ‘precise’. Is using the clock’s stamp insufficiently accurate?

**Comment by gennartan** [comment]
> I think an illustrative example of what you’re looking to do might help me give a suggestion. I’m not sure that should be required here, but maybe we’re thinking about different definitions of ‘precise’. Is using the clock’s stamp insufficiently accurate?

My question is more about the clock control in the test. If I want to test if the polygon action is correctly applied after 10s of positive detection, I don't want to wait for 10 actual seconds on the system running the tests using a sleep function, I want to be able to change the clock vlaue to simulate those 10 seconds. 

Do you know if such a function exists ?

**Comment by SteveMacenski** [comment]
> I don't want to wait for 10 actual seconds on the system running the tests using a sleep function, I want to be able to change the clock value to simulate those 10 seconds.

You could set simulation time and have your own `/clock` publisher, but that seems like overkill to bake into the test. I suggest making some of the times smaller values like 2-3 seconds instead. Keep the test times down whereever reasonably possible, but don't go overboard.

**Comment by SteveMacenski** [comment]
Hi @gennartan how can I help push this along? :smile:

**Comment by gennartan** [comment]
Hi, sorry, I am a bit blocked :smile: 
My main concern is that I don't want that feature to cause bugs, and I recently had a few bugs with time jumps with my current system causing some nodes to bug silently.

I will soon push my latest updates of the code (I fixed the tests that were failing), but I think I still need a bit more work on this to be sure not to create bugs

**Comment by SteveMacenski** [comment]
Ok, I appreciate your caution!

**Comment by SteveMacenski** [comment]
Ok! thanks for letting us know!
